File: day01/main.zig
const std = @import("std");

const input = @embedFile("./input.txt");
// const input =
//     \\two1nine
//     \\eightwothree
//     \\abcone2threexyz
//     \\xtwone3four
//     \\4nineeightseven2
//     \\zoneight234
//     \\7pqrstsixteen
// ;

pub fn main() !void {
    std.debug.print("{d}\n", .{getSumOfCalibration(input)});
}

fn getSumOfCalibration(buffer: []const u8) u32 {
    var lines = std.mem.split(u8, buffer, "\n");
    var total: u32 = 0;
    while (lines.next()) |line| {
        const lineScore = part2(line);
        std.debug.print("{d}\n", .{lineScore});
        total += lineScore;
    }

    return total;
}

fn part2(line: []const u8) u8 {
    const startsWith = std.mem.startsWith;

    var first: ?u8 = null;
    var last: ?u8 = null;
    var idx: u32 = 0;

    while (true) {
        if (idx >= line.len)
            break;

        switch (line[idx]) {
            '0'...'9' => {
                if (first) |_| {} else {
                    first = line[idx] - 0x30;
                }
                last = line[idx] - 0x30;
                idx += 1;
            },
            'o' => {
                if (startsWith(u8, line[idx..line.len], "one")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 1;
                    }

                    last = 1;
                } else {
                    idx += 1;
                }
            },
            't' => {
                if (startsWith(u8, line[idx..line.len], "two")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 2;
                    }

                    last = 2;
                } else if (startsWith(u8, line[idx..line.len], "three")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 3;
                    }
                    last = 3;
                } else {
                    idx += 1;
                }
            },
            'f' => {
                if (startsWith(u8, line[idx..line.len], "four")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 4;
                    }

                    last = 4;
                } else if (startsWith(u8, line[idx..line.len], "five")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 5;
                    }
                    last = 5;
                } else {
                    idx += 1;
                }
            },
            's' => {
                if (startsWith(u8, line[idx..line.len], "six")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 6;
                    }

                    last = 6;
                } else if (startsWith(u8, line[idx..line.len], "seven")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 7;
                    }
                    last = 7;
                } else {
                    idx += 1;
                }
            },
            'e' => {
                if (startsWith(u8, line[idx..line.len], "eight")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 8;
                    }

                    last = 8;
                } else {
                    idx += 1;
                }
            },
            'n' => {
                if (startsWith(u8, line[idx..line.len], "nine")) {
                    idx += 1;
                    if (first) |_| {} else {
                        first = 9;
                    }

                    last = 9;
                } else {
                    idx += 1;
                }
            },

            else => idx += 1,
        }
    }
    return (first orelse 0) * 10 + (last orelse 0);
}

/// part 1
fn getFirstAndLastNumber(line: []const u8) u8 {
    var first: ?u8 = null;
    var last: ?u8 = null;

    for (line) |char| {
        switch (char) {
            '0'...'9' => {
                if (first) |_| {} else {
                    first = char - 0x30;
                }

                last = char - 0x30;
            },
            else => {},
        }
    }
    return (first orelse 0) * 10 + (last orelse 0);
}

File: day01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <memory.h>

#define SUCCESS 0
#define FAILURE 1
typedef unsigned long u64;

static FILE *fp = NULL;
static char *buffer = NULL;
u64 length = 0;

static u64 converge_left(char *);
static u64 converge_right(char *);
static int number_from(char const *);

char const *NUMBERS[] = {
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
};

int main(void) {
    int STATUS = SUCCESS;

    fp = fopen("day01/input.txt", "r");
    if (!fp) {
        STATUS = FAILURE;
        fprintf(stderr, "Could not open the file\n");
        goto cleanup;
    }
    fseek(fp, 0, SEEK_END);
    length = ftell(fp);
    rewind(fp);

    if (!length) {
        STATUS = FAILURE;
        fprintf(stderr, "There is nothing in the file\n");
        goto cleanup;
    }

    buffer = (char *)malloc(length * sizeof(char));
    if (!buffer) {
        STATUS = FAILURE;
        fprintf(stderr, "Could not allocate a buffer\n");
        goto cleanup;
    }

    if (!fread(buffer, sizeof(char), length, fp)) {
        STATUS = FAILURE;
        fprintf(stderr, "Read nothing from the file\n");
        goto cleanup;
    }

    u64 total_part2 = 0;
    char *current = buffer;

    while (1) {
        char *next = strstr(current, "\n");
        if (!next)
            next = buffer + length;
        next -= 1;

        int line_total = converge_left(current) + converge_right(next);

        total_part2 += line_total;
        current = next + 2;

        if (current >= buffer + length)
            break;
    }

    printf("answer part 2: %ld\n", total_part2);

cleanup:
    if (buffer)
        free(buffer);

    if (fp)
        fclose(fp);

    return STATUS;
}

static u64 converge_left(char *line) {
    for (; *line != '\n'; ++line) {
        u64 number = number_from(line);
        if (number)
            return number * 10;
    }
    // this should never happen, but we dont want implicit return, hidden flow
    return 0;
}

static u64 converge_right(char *line) {
    for (; *line != '\n'; --line) {
        u64 number = number_from(line);
        if (number)
            return number;
    }
    return 0;
}

static int number_from(char const *line) {
    if (*line >= '1' && *line <= '9')
        return *line - '0';

    // 9 numbers, so we hard code it
    for (int i = 0; i < 9; ++i) {
        if (!strncmp(line, NUMBERS[i], strlen(NUMBERS[i])))
            return i + 1;
    }
    return 0;
}